src/restruct_poc/model/model.hpp

```diff
@@ -15,64 +16,138 @@ static Node InvalidNode(-1, {0., 0., 0., 1., 0., 0., 0.});
 /// relationships between components in a turbine. Nodes represent points in space with
 /// position, displacement, velocity, and acceleration. Constraints represent relationships
 /// between nodes that restrict their relative motion in some way.
-struct Model {
-    std::vector<Node> nodes;
-    std::vector<Constraint> constraints;
+class Model {
+public:
+    Model() = default;
 
-    /// Adds a node to the model and returns the node
-    Node AddNode(
+    Model(std::vector<Node> nodes, std::vector<BeamElement> beam_elements) {
+        for (auto& node : nodes) {
+            this->nodes_.push_back(std::make_shared<Node>(node));
+        }
+        for (auto& element : beam_elements) {
+            this->beam_elements_.push_back(std::make_shared<BeamElement>(element));
+        }
+    }
+
+    Model(
+        std::vector<std::shared_ptr<Node>> nodes,
+        std::vector<std::shared_ptr<BeamElement>> beam_elements
+    )
+        : nodes_(std::move(nodes)), beam_elements_(std::move(beam_elements)) {}
+
+    // Make the class non-copyable and non-assignable
+    Model(const Model&) = delete;
+    Model& operator=(const Model&) = delete;
```

Comment:
Why do these need to be deleted?  I would propose allowing them for shallow copies and adding a "CloneModel" function (or similar name) for when we want a deep copy.

Reply:
My opinion is we don't want to make copies of the `Model` class objects (either shallow or deep) since it's a top-level data structure containing lots of resources. Should we wait for a use case to arise before we permit this?

Reply:
I'm of the opposite persuasion - unless we have good reason, we shouldn't touch the copy mechanics unless we have a specific reason to need to disable them.  Deleting them here also gets rid of things like the move constructor.  See: https://en.cppreference.com/w/cpp/language/rule_of_three

Reply:
While I don't see any use case for copy/move operations on the `Model` class, let's just follow rule of zero here.